### ISSUE TITLE
[get.jenkins.io] provide a mirror for Jenkins Russian users

### ISSUE THREAD

**Issue by dduportal** [jenkins-infra/helpdesk#4147]
### Service(s)

get.jenkins.io

### Summary

As per https://github.com/jenkins-infra/helpdesk/issues/4128#issuecomment-2160116396, the Jenkins Russian users recently went trough troubles when it comes to downloading Jenkins plugins/installers through our mirror system `get.jenkins.io`:

- The closest mirror, for Russian users, used to be [Aachen University}(https://ftp.halifax.rwth-aachen.de/jenkins/). It recently has been blocked by [Roskomnadzor](https://en.wikipedia.org/wiki/Roskomnadzor) for reasons outside the Jenkins project scope (and ability to act).
- Hopefully, we recently benefited from a new mirror in Romania thanks to [Hostico](https://github.com/jenkins-infra/helpdesk/issues/3976) (see our [announcement](https://www.jenkins.io/blog/2024/06/13/thanks-mirror-sponsors/))
- Alas, our friends at Hostico are now receiving a LOT of traffic due to their proximity which is not sustainable at all

As such, this issue is to find a solution (and track its resolution up to completion) from the Jenkins project infrastructure.
The goals are:

* Ensure Hostico can keep sponsoring us sustainably by keeping their traffic under an acceptable (for them!) threshold
* Ensure that Jenkins Russian users have a download mirror, not blocked with acceptable performances
* Ensure that the Jenkins Infrastructure can sustains the cost of any solution here (sponsored, credits, billin, any kind)

**Comment by dduportal** [comment]
First set of options after initial discussions:

- It looks like there is an [unofficial Yandex hosted mirror](https://github.com/jenkins-infra/helpdesk/issues/4128#issuecomment-2160320848) at https://mirror.yandex.ru/mirrors/jenkins/plugins/ => let's contact them to see if we can "simply" (modulo having an rsync or FTP service for scanning). Could be the easiest solution!
- The Jenkins project currently has access to different clouds with sponsored credits (or billing) where a mirror could be managed by the Jenkins infra team with a location close to Russia (at least West Russia) for QoS, and we could set up fake coordinates in the mirror setup inside `mirrorbits` to force this mirror to be picked when requests comes from West Russia (instead of fall backing to Hostico in Romania)
  - we have an AWS account. Since AWS has DC in Sweden (https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Concepts.RegionsAndAvailabilityZones.html), it could be close enough
  - Azure has a Poland region: https://azure.microsoft.com/en-us/updates/general-availability-microsoft-azure-available-from-new-cloud-region-in-poland/
  - DigitalOcean has a Frankfurt region from where we operate
  - We could also try to start new sponsor accounts: OVH has Poland DCs as well (see https://www.ovhcloud.com/en/vps/vps-poland/)

**Comment by mikhirev** [comment]
I tried to contact the mirror.yandex team via the official email address, but got no reply. Maybe @w-e-g could directly contact the engineer he mentioned.

**Comment by w-e-g** [comment]
My friend pinged a man who worked on mirror.yandex.ru, but no answer at all.

**Comment by timja** [comment]
We've had a response from yandex to the jenkins infra team FYI, they would like to make their mirror an official one.

**Comment by dduportal** [comment]
We're going to add this mirror to get.jenkins.io today (Monday 24 June 2024) between 01:35pm UTC and 02:00pm UTC (along with https://github.com/jenkins-infra/helpdesk/issues/4145)

**Comment by lemeurherve** [comment]
Yandex mirror has been added with success 🥳

https://get.jenkins.io/debian/jenkins_2.463_all.deb?mirrorstats

Note: we had the same issue as for Hostico (mirror marked as "down", cf https://github.com/jenkins-infra/helpdesk/issues/3976#issuecomment-2158466326), resolved after some times.

Not sure what worked, `mirrorbits edit` wasn't at first, we then deleted and readded the mirror but it was only after seeing files being individually scanned in mirrorbits logs that the mirror appeared as "up".

**Comment by dduportal** [comment]
> yandex is blocked in a lot of corporate environments.

Hi @Unknown-Guy , there isn't a lot we can do on the infra, unless one of these corporations helps by providing a (public) download mirror in the Russia area. As a reminder, the Jenkins project is living on sponsorships: we do not have the financial resource to cover all cases and quality of service unless users helps by sponsoring.

Other alternative is to contact companies such as CloudBees (and pay their product) to ensure this kind of constraint is solved for your company.

**Comment by Stikus** [comment]
Looks like this is vice versa situation.

@dduportal It is time to add mirror selection preferences on Jenkins end, what do you think?

**Comment by lemeurherve** [comment]
`CountryOnly` is already set to `RU` in the mirrorbits configuration of Yandex mirror.

**Comment by Stikus** [comment]
@QMiqTx6DHn1bA9yaNaAbsD3CLG8gTmd4 This addition will fix current problem (for some EU countries), but will lead for new problems:

- yandex mirror will still be best for some countries, but they cannot reach it due to `CountryCodes` blocking
- hostico traffic will increase and this was a problem (as stated in the beginning of the topic)
- some other mirrors can be blocked for various reasons - it has to be setting to add preferable mirror for Jenkins server admin

I think we need another issue for this discussion, what do you think?

**Comment by dduportal** [comment]
Hi folks, thanks for the pointers and details.

- @Unknown-Guy thanks for commenting and detailing after my comment that you are in Europe (but close to the Yandex mirror). This is not a breaking change because this is an open source project living by sponsoring and there are no rules of engagement as every contributor is doing their best. Don't get me wrong: we understand and acknowledge the issue now you've pointed that you are in EU (I did not know you were with your first message): we are going to do something but ETA is next week. We are sorry for the inconvenience.
- @QMiqTx6DHn1bA9yaNaAbsD3CLG8gTmd4 thanks for the reminder. As @lemeurherve proposed, we should be able to act on this next week, but we need to plan it carefully to avoid problems such as described in https://github.com/jenkins-infra/helpdesk/issues/4147#issuecomment-2196583357

>  is time to add mirror selection preferences on Jenkins end, what do you think?

If this is something which look interesting to you, you can raise a feature request on [mirrorbits](https://github.com/etix/mirrorbits) open source software we use: if they implement it then we could benefit from it. However, I don't think it is a wise choice as it would allow anyone to bypass the redirections, which defeat the whole purpose (ref. https://www.sonatype.com/blog/maven-central-and-the-tragedy-of-the-commons). If the goal is to solve local optimums and break the other, then it start to be a problem. 
If you have a solution which does not involve weeks of works for the infra contributors, please share it with us we are all hears!

**Comment by dduportal** [comment]
Reopening as this is a problem to be solved

**Comment by dduportal** [comment]
Proposal based on the above comments from different users + internal discussions:

- Set Yandex mirror to Russian country only => this mirror would only be selected for Russian users
- All East-Europe traffic will get to Hostico due to this one => if is not sustainable for them, then we'll also restrict them to Romania only
- It means, in the worst case scenario, we'll back to the situation one month ago with traffic going to Germany and/or Belgium (except for Russian and Romania) so still better than before


Any objection of thing to consider if we set up this change next week?

**Comment by Stikus** [comment]
@dduportal Can you describe what Jenkins will do if first mirror picked is not reachable for some reason? Will it fail or pick second mirror? If it can pick second mirror - it will fix all these problems, don't you think so? Or it is hard to implement? (I'm not familiar with mirrorbits, sorry)

**Comment by dduportal** [comment]
> @dduportal Can you describe what Jenkins will do if first mirror picked is not reachable for some reason? Will it fail or pick second mirror? If it can pick second mirror - it will fix all these problems, don't you think so? Or it is hard to implement? (I'm not familiar with mirrorbits, sorry)

No problem, this is a good question. Mirrorbits only emits an HTTP 3xx redirection, with the Location header set to the mirror URL closest to you. There is no dynamic reverse proxying so we cannot do such feature.

the goal for it is to be simple to defer traffic to mirrors (handling only http/3xx request does not cost much). So it can handle a lot of traffic.

**Comment by dduportal** [comment]
> Proposal based on the above comments from different users + internal discussions:
> 
>     * Set Yandex mirror to Russian country only => this mirror would only be selected for Russian users
> 
>     * All East-Europe traffic will get to Hostico due to this one => if is not sustainable for them, then we'll also restrict them to Romania only
> 
>     * It means, in the worst case scenario, we'll back to the situation one month ago with traffic going to Germany and/or Belgium (except for Russian and Romania) so still better than before
> 
> 
> Any objection of thing to consider if we set up this change next week?

We are starting the operation

**Comment by dduportal** [comment]
> > Proposal based on the above comments from different users + internal discussions:
> > ```
> > * Set Yandex mirror to Russian country only => this mirror would only be selected for Russian users
> > 
> > * All East-Europe traffic will get to Hostico due to this one => if is not sustainable for them, then we'll also restrict them to Romania only
> > 
> > * It means, in the worst case scenario, we'll back to the situation one month ago with traffic going to Germany and/or Belgium (except for Russian and Romania) so still better than before
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any objection of thing to consider if we set up this change next week?
> 
> We are starting the operation

Operation finished: both Yandex and Hostico mirrors are now set up as "Country only".
It means than these mirrors will only be selected to serve traffic originating from the country on which they are marked from (Russia and Romania).

You can see this result in https://get.jenkins.io/war/2.464/jenkins.war?mirrorlist as example:

<img width="1821" alt="Capture d’écran 2024-07-01 à 12 14 26" src="https://github.com/jenkins-infra/helpdesk/assets/1522731/32766ed5-2423-4017-8fb3-983f3f0b56c9">

**Comment by dduportal** [comment]
Ping @anonimus-asuseeepc (ref. https://github.com/jenkins-infra/helpdesk/issues/4160) @QMiqTx6DHn1bA9yaNaAbsD3CLG8gTmd4 @Stikus @Unknown-Guy could you confirm if it fixes your problems?
